src/cli/term2.rs

```diff
@@ -3,238 +3,150 @@
 //! if TERM isn't defined.
 
 use std::io;
-use std::ops::Deref;
-use std::sync::Mutex;
+use std::io::Write;
 
-use lazy_static::lazy_static;
-pub use term::color;
-pub use term::Attr;
-pub use term::Terminal;
+use termcolor::{ColorChoice, ColorSpec, StandardStream, WriteColor};
 
-use crate::currentprocess::filesource::{Isatty, Writer};
 use crate::process;
 
-mod termhack {
-    // Things we should submit to term as improvements: here temporarily.
-    use std::collections::HashMap;
-    use std::io;
-
-    use term::terminfo::TermInfo;
-    use term::{StderrTerminal, StdoutTerminal, Terminal, TerminfoTerminal};
-
-    /// Return a Terminal object for T on this platform.
-    /// If there is no terminfo and the platform requires terminfo, then None is returned.
-    fn make_terminal<T, F>(
-        terminfo: Option<TermInfo>,
-        source: F,
-    ) -> Option<Box<dyn Terminal<Output = T> + Send>>
-    where
-        T: 'static + io::Write + Send,
-        // Works around stdio instances being unclonable.
-        F: Fn() -> T + Copy,
-    {
-        let result = terminfo
-            .map(move |ti| TerminfoTerminal::new_with_terminfo(source(), ti))
-            .map(|t| Box::new(t) as Box<dyn Terminal<Output = T> + Send>);
-        #[cfg(windows)]
-        {
-            result.or_else(|| {
-                term::WinConsole::new(source())
-                    .ok()
-                    .map(|t| Box::new(t) as Box<dyn Terminal<Output = T> + Send>)
-            })
-        }
-        #[cfg(not(windows))]
-        {
-            result
-        }
-    }
+pub trait Terminal: io::Write {
+    fn fg(&mut self, color: Color) -> io::Result<()>;
+    fn bg(&mut self, color: Color) -> io::Result<()>;
+    fn attr(&mut self, attr: Attr) -> io::Result<()>;
+    fn reset(&mut self) -> io::Result<()>;
+    fn carriage_return(&mut self) -> io::Result<()>;
+}
 
-    pub fn make_terminal_with_fallback<T, F>(
-        terminfo: Option<TermInfo>,
-        source: F,
-    ) -> Box<dyn Terminal<Output = T> + Send>
-    where
-        T: 'static + io::Write + Send,
-        // Works around stdio instances being unclonable.
-        F: Fn() -> T + Copy,
-    {
-        make_terminal(terminfo, source)
-            .or_else(|| {
-                let ti = TermInfo {
-                    names: vec![],
-                    bools: HashMap::new(),
-                    numbers: HashMap::new(),
-                    strings: HashMap::new(),
-                };
-                let t = TerminfoTerminal::new_with_terminfo(source(), ti);
-                Some(Box::new(t) as Box<dyn Terminal<Output = T> + Send>)
-            })
-            .unwrap()
-    }
-    /// Return a Terminal wrapping stdout, or None if a terminal couldn't be
-    /// opened.
-    #[allow(unused)]
-    pub fn stdout(terminfo: Option<TermInfo>) -> Option<Box<StdoutTerminal>> {
-        make_terminal(terminfo, io::stdout)
-    }
+#[derive(Copy, Clone, Debug)]
+pub enum Color {
```

Comment:
The intent was to encapsulate all terminal related things in `term2` so that future refactor or swapping of `termcolor` for something else has minimal impact.

Reply:
Currently rustup doesn't offer a crate; so we can simply hit all code paths that are affected in a single commit like you are if a future change is made. I'd like @kinnison's opinion too, but my view today is that we value less code over future possibilities.